storagenode/nodestats/service.go

```diff
@@ -66,21 +43,131 @@ func (c *Client) Close() error {
 type Service struct {
 	log *zap.Logger
 
+	statsTicker *time.Ticker
+	spaceTicker *time.Ticker
+
 	transport transport.Client
+	consoleDB console.DB
 	kademlia  *kademlia.Kademlia
 }
 
 // NewService creates new instance of service
-func NewService(log *zap.Logger, transport transport.Client, kademlia *kademlia.Kademlia) *Service {
+func NewService(log *zap.Logger, transport transport.Client, consoleDB console.DB, kademlia *kademlia.Kademlia) *Service {
 	return &Service{
-		log:       log,
-		transport: transport,
-		kademlia:  kademlia,
+		log:         log,
+		statsTicker: time.NewTicker(time.Hour * 4),
+		spaceTicker: time.NewTicker(time.Hour * 12),
+		transport:   transport,
+		consoleDB:   consoleDB,
+		kademlia:    kademlia,
+	}
+}
+
+// Run runs loop
+func (s *Service) Run(ctx context.Context) (err error) {
+	err = s.CacheStatsFromSatellites(ctx)
+	if err != nil {
+		s.log.Error(fmt.Sprintf("Get stats query failed: %v", err))
+	}
+
+	err = s.CacheSpaceUsageFromSatellites(ctx)
+	if err != nil {
+		s.log.Error(fmt.Sprintf("Get disk space usage query failed: %v", err))
+	}
+
+	for {
+		select {
+		// handle cancellation signal first
+		case <-ctx.Done():
+			return ctx.Err()
+		// wait for the next stats interval to happen
+		case <-s.statsTicker.C:
+			err = s.CacheStatsFromSatellites(ctx)
+			if err != nil {
+				s.log.Error(fmt.Sprintf("Get stats query failed: %v", err))
+			}
+		// wait for the next space interval to happen
+		case <-s.spaceTicker.C:
+			err = s.CacheSpaceUsageFromSatellites(ctx)
+			if err != nil {
+				s.log.Error(fmt.Sprintf("Get disk space usage query failed: %v", err))
+			}
+		}
 	}
 }
 
+// CacheStatsFromSatellites queries node stats from all the satellites
+// known to the storagenode and stores this information into db
+func (s *Service) CacheStatsFromSatellites(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	satellites, err := s.consoleDB.Satellites().GetIDs(ctx)
+	if err != nil {
+		return NodeStatsServiceErr.Wrap(err)
+	}
+
+	var cacheStatsErr errs.Group
+	for _, satellite := range satellites {
+		stats, err := s.GetStatsFromSatellite(ctx, satellite)
+		if err != nil {
+			cacheStatsErr.Add(err)
+			continue
+		}
+
+		stats.UpdatedAt = time.Now()
+
+		// try to update stats from satellite
+		if err = s.consoleDB.Stats().Update(ctx, *stats); err != nil {
+			// if stats doesn't exists - create new one
+			if err == sql.ErrNoRows {
+				_, err = s.consoleDB.Stats().Create(ctx, *stats)
```

Comment:
This should happen as a single update not separately. This kind of code is inherently racy and will cause problems.